File: task4.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>

using namespace std;

// Функция для вывода доски
void printBoard(const vector<vector<int>>& board, int n, int m) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cout << board[i][j];
            if (j < m - 1) cout << " ";
        }
        cout << endl;
    }
}

// Функция для подсчета ходов (без вывода шагов)
int countMovesToHalf(const vector<vector<int>>& board, int n, int m, 
                     int topValue, int bottomValue) {
    int moves = 0;
    int midRow = n / 2;
    vector<vector<int>> temp = board;
    
    // Обрабатываем верхнюю половину
    for (int i = 0; i < midRow; i++) {
        for (int j = 0; j < m; j++) {
            if (temp[i][j] != topValue) {
                int dx[] = {0, 1, 0, -1};
                int dy[] = {1, 0, -1, 0};
                
                for (int k = 0; k < 4; k++) {
                    int ni = i + dx[k];
                    int nj = j + dy[k];
                    
                    if (ni >= 0 && ni < n && nj >= 0 && nj < m) {
                        temp[i][j] = 1 - temp[i][j];
                        temp[ni][nj] = 1 - temp[ni][nj];
                        moves++;
                        break;
                    }
                }
            }
        }
    }
    
    // Обрабатываем нижнюю половину
    for (int i = midRow; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (temp[i][j] != bottomValue) {
                int dx[] = {0, 1, 0, -1};
                int dy[] = {1, 0, -1, 0};
                
                for (int k = 0; k < 4; k++) {
                    int ni = i + dx[k];
                    int nj = j + dy[k];
                    
                    if (ni >= 0 && ni < n && nj >= 0 && nj < m) {
                        temp[i][j] = 1 - temp[i][j];
                        temp[ni][nj] = 1 - temp[ni][nj];
                        moves++;
                        break;
                    }
                }
            }
        }
    }
    
    return moves;
}

// Функция для выполнения решения с выводом шагов
void solveWithSteps(const vector<vector<int>>& board, int n, int m, 
                    int topValue, int bottomValue) {
    int moves = 0;
    int midRow = n / 2;
    vector<vector<int>> temp = board;
    
    cout << "\nЦелевая конфигурация: верх=" << topValue << ", низ=" << bottomValue << "\n";
    cout << "Начальное состояние:\n";
    printBoard(temp, n, m);
    
    // Верхняя половина
    for (int i = 0; i < midRow; i++) {
        for (int j = 0; j < m; j++) {
            if (temp[i][j] != topValue) {
                int dx[] = {0, 1, 0, -1};
                int dy[] = {1, 0, -1, 0};
                string dir[] = {"вправо", "вниз", "влево", "вверх"};
                
                for (int k = 0; k < 4; k++) {
                    int ni = i + dx[k];
                    int nj = j + dy[k];
                    
                    if (ni >= 0 && ni < n && nj >= 0 && nj < m) {
                        moves++;
                        cout << "\nХод #" << moves << ": Переворачиваем (" 
                             << i << "," << j << ") и (" << ni << "," << nj << ") [" << dir[k] << "]\n";
                        
                        temp[i][j] = 1 - temp[i][j];
                        temp[ni][nj] = 1 - temp[ni][nj];
                        
                        cout << "  После: (" << i << "," << j << ")=" << temp[i][j] 
                             << ", (" << ni << "," << nj << ")=" << temp[ni][nj] << "\n";
                        cout << "Текущее состояние:\n";
                        printBoard(temp, n, m);
                        break;
                    }
                }
            }
        }
    }
    
    // Нижняя половина
    for (int i = midRow; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (temp[i][j] != bottomValue) {
                int dx[] = {0, 1, 0, -1};
                int dy[] = {1, 0, -1, 0};
                string dir[] = {"вправо", "вниз", "влево", "вверх"};
                
                for (int k = 0; k < 4; k++) {
                    int ni = i + dx[k];
                    int nj = j + dy[k];
                    
                    if (ni >= 0 && ni < n && nj >= 0 && nj < m) {
                        moves++;
                        cout << "\nХод #" << moves << ": Переворачиваем (" 
                             << i << "," << j << ") и (" << ni << "," << nj << ") [" << dir[k] << "]\n";
                        
                        temp[i][j] = 1 - temp[i][j];
                        temp[ni][nj] = 1 - temp[ni][nj];
                        
                        cout << "  После: (" << i << "," << j << ")=" << temp[i][j] 
                             << ", (" << ni << "," << nj << ")=" << temp[ni][nj] << "\n";
                        cout << "Текущее состояние:\n";
                        printBoard(temp, n, m);
                        break;
                    }
                }
            }
        }
    }
    
    cout << "\nВсего ходов: " << moves << "\n";
}

// Генерация случайной доски
void generateRandomBoard(vector<vector<int>>& board, int n, int m) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            board[i][j] = rand() % 2;
        }
    }
}

int main() {
    setlocale(LC_ALL, "rus");
    srand(time(0));
    
    int n, m;
    cout << "Введите количество строк (N): ";
    cin >> n;
    cout << "Введите количество столбцов (M): ";
    cin >> m;
    
    vector<vector<int>> board(n, vector<int>(m));
    
    // Выбор способа ввода
    cout << "\nВыберите способ заполнения доски:\n";
    cout << "1. Ручной ввод\n";
    cout << "2. Случайная генерация\n";
    cout << "Ваш выбор: ";
    int choice;
    cin >> choice;
    
    if (choice == 1) {
        cout << "\nВведите состояние доски (0 - орел, 1 - решка):\n";
        for (int i = 0; i < n; i++) {
            cout << "Строка " << i << ": ";
            for (int j = 0; j < m; j++) {
                cin >> board[i][j];
            }
        }
    } else {
        generateRandomBoard(board, n, m);
    }
    
    cout << "\nСгенерированная доска:\n";
    printBoard(board, n, m);
    
    // Считаем ходы для обоих вариантов
    int moves1 = countMovesToHalf(board, n, m, 0, 1);
    int moves2 = countMovesToHalf(board, n, m, 1, 0);
    
    // Определяем оптимальный вариант
    int bestTop, bestBottom;
    if (moves1 <= moves2) {
        bestTop = 0;
        bestBottom = 1;
    } else {
        bestTop = 1;
        bestBottom = 0;
    }
    
    // Выбор режима отображения шагов
    cout << "\nПоказывать пошаговое решение? (1 - да, 0 - нет): ";
    int showStepsChoice;
    cin >> showStepsChoice;
    
    if (showStepsChoice == 1) {
        // Выводим шаги только для оптимального варианта
        solveWithSteps(board, n, m, bestTop, bestBottom);
    }
    
    // Итоговый результат
    cout << "\n Итоги\n";
    cout << "Минимальное количество ходов: " << min(moves1, moves2) << "\n";
    cout << "Оптимальная конфигурация: верхняя половина - " << bestTop 
         << ", нижняя - " << bestBottom << "\n";
    
    return 0;
}

File: task4.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// Функция для вывода доски
func printBoard(board [][]int, n, m int) {
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			fmt.Print(board[i][j])
			if j < m-1 {
				fmt.Print(" ")
			}
		}
		fmt.Println()
	}
}

// Функция для подсчета ходов (без вывода шагов)
func countMovesToHalf(board [][]int, n, m, topValue, bottomValue int) int {
	moves := 0
	midRow := n / 2
	
	// Глубокое копирование доски (слайсы в Го — ссылочный тип)
	temp := make([][]int, n)
	for i := range temp {
		temp[i] = make([]int, m)
		copy(temp[i], board[i])
	}
	
	// Обрабатываем верхнюю половину
	for i := 0; i < midRow; i++ {
		for j := 0; j < m; j++ {
			if temp[i][j] != topValue {
				dx := []int{0, 1, 0, -1}
				dy := []int{1, 0, -1, 0}
				
				for k := 0; k < 4; k++ {
					ni := i + dx[k]
					nj := j + dy[k]
					
					if ni >= 0 && ni < n && nj >= 0 && nj < m {
						temp[i][j] = 1 - temp[i][j]
						temp[ni][nj] = 1 - temp[ni][nj]
						moves++
						break
					}
				}
			}
		}
	}
	
	// Обрабатываем нижнюю половину
	for i := midRow; i < n; i++ {
		for j := 0; j < m; j++ {
			if temp[i][j] != bottomValue {
				dx := []int{0, 1, 0, -1}
				dy := []int{1, 0, -1, 0}
				
				for k := 0; k < 4; k++ {
					ni := i + dx[k]
					nj := j + dy[k]
					
					if ni >= 0 && ni < n && nj >= 0 && nj < m {
						temp[i][j] = 1 - temp[i][j]
						temp[ni][nj] = 1 - temp[ni][nj]
						moves++
						break
					}
				}
			}
		}
	}
	
	return moves
}

// Функция для выполнения решения с выводом шагов
func solveWithSteps(board [][]int, n, m, topValue, bottomValue int) {
	moves := 0
	midRow := n / 2
	
	// Глубокое копирование доски
	temp := make([][]int, n)
	for i := range temp {
		temp[i] = make([]int, m)
		copy(temp[i], board[i])
	}
	
	fmt.Printf("\nЦелевая конфигурация: верх=%d, низ=%d\n", topValue, bottomValue)
	fmt.Println("Начальное состояние:")
	printBoard(temp, n, m)
	
	// Направления и их названия
	dx := []int{0, 1, 0, -1}
	dy := []int{1, 0, -1, 0}
	dir := []string{"вправо", "вниз", "влево", "вверх"}
	
	// Верхняя половина
	for i := 0; i < midRow; i++ {
		for j := 0; j < m; j++ {
			if temp[i][j] != topValue {
				for k := 0; k < 4; k++ {
					ni := i + dx[k]
					nj := j + dy[k]
					
					if ni >= 0 && ni < n && nj >= 0 && nj < m {
						moves++
						fmt.Printf("\nХод #%d: Переворачиваем (%d,%d) и (%d,%d) [%s]\n",
							moves, i, j, ni, nj, dir[k])
						
						temp[i][j] = 1 - temp[i][j]
						temp[ni][nj] = 1 - temp[ni][nj]
						
						fmt.Printf("  После: (%d,%d)=%d, (%d,%d)=%d\n",
							i, j, temp[i][j], ni, nj, temp[ni][nj])
						fmt.Println("Текущее состояние:")
						printBoard(temp, n, m)
						break
					}
				}
			}
		}
	}
	
	// Нижняя половина
	for i := midRow; i < n; i++ {
		for j := 0; j < m; j++ {
			if temp[i][j] != bottomValue {
				for k := 0; k < 4; k++ {
					ni := i + dx[k]
					nj := j + dy[k]
					
					if ni >= 0 && ni < n && nj >= 0 && nj < m {
						moves++
						fmt.Printf("\nХод #%d: Переворачиваем (%d,%d) и (%d,%d) [%s]\n",
							moves, i, j, ni, nj, dir[k])
						
						temp[i][j] = 1 - temp[i][j]
						temp[ni][nj] = 1 - temp[ni][nj]
						
						fmt.Printf("  После: (%d,%d)=%d, (%d,%d)=%d\n",
							i, j, temp[i][j], ni, nj, temp[ni][nj])
						fmt.Println("Текущее состояние:")
						printBoard(temp, n, m)
						break
					}
				}
			}
		}
	}
	
	fmt.Printf("\nВсего ходов: %d\n", moves)
}

// Генерация случайной доски
func generateRandomBoard(board [][]int, n, m int) {
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			board[i][j] = rand.Intn(2)
		}
	}
}

// Вспомогательная функция min для целых чисел
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	// Инициализация генератора случайных чисел
	rand.Seed(time.Now().UnixNano())
	
	var n, m int
	fmt.Print("Введите количество строк (N): ")
	fmt.Scan(&n)
	fmt.Print("Введите количество столбцов (M): ")
	fmt.Scan(&m)
	
	// Создание доски
	board := make([][]int, n)
	for i := range board {
		board[i] = make([]int, m)
	}
	
	// Выбор способа ввода
	fmt.Println("\nВыберите способ заполнения доски:")
	fmt.Println("1. Ручной ввод")
	fmt.Println("2. Случайная генерация")
	fmt.Print("Ваш выбор: ")
	var choice int
	fmt.Scan(&choice)
	
	if choice == 1 {
		fmt.Println("\nВведите состояние доски (0 - орел, 1 - решка):")
		for i := 0; i < n; i++ {
			fmt.Printf("Строка %d: ", i)
			for j := 0; j < m; j++ {
				fmt.Scan(&board[i][j])
			}
		}
	} else {
		generateRandomBoard(board, n, m)
	}
	
	fmt.Println("\nСгенерированная доска:")
	printBoard(board, n, m)
	
	// Считаем ходы для обоих вариантов
	moves1 := countMovesToHalf(board, n, m, 0, 1)
	moves2 := countMovesToHalf(board, n, m, 1, 0)
	
	// Определяем оптимальный вариант
	bestTop, bestBottom := 0, 1
	if moves1 > moves2 {
		bestTop, bestBottom = 1, 0
	}
	
	// Выбор режима отображения шагов
	fmt.Print("\nПоказывать пошаговое решение? (1 - да, 0 - нет): ")
	var showStepsChoice int
	fmt.Scan(&showStepsChoice)
	
	if showStepsChoice == 1 {
		// Выводим шаги только для оптимального варианта
		solveWithSteps(board, n, m, bestTop, bestBottom)
	}
	
	// Итоговый результат
	fmt.Println("\n Итоги")
	fmt.Printf("Минимальное количество ходов: %d\n", min(moves1, moves2))
	fmt.Printf("Оптимальная конфигурация: верхняя половина - %d, нижняя - %d\n",
		bestTop, bestBottom)
}